File: blind75/twoPointers/15-3Sum.js
// 15. 3Sum
// Given an integer array nums, return all the triplets [nums[i], nums[j], nums[k]] such that i != j, i != k, and j != k, and nums[i] + nums[j] + nums[k] == 0.
// Notice that the solution set must not contain duplicate triplets.

/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    let res = [];
    nums.sort(function(a, b){return a - b});
    let len = nums.length;
    for (let i = 0; i < len; i++) {
        // when the value is same as the before value, we need to avoid this case
        if (i > 0 && nums[i] === nums[i - 1])
            continue;
        let l = i + 1, r = len - 1;
        while(l < r) {
            let sum = nums[i] + nums[l] + nums[r];
            if (sum > 0)
                r--;
            else if (sum < 0)
                l++;
            else {
                res.push([nums[i], nums[l], nums[r]]);
                l++;
                // for avoiding adding the same array
                while(l < r && nums[l - 1] === nums[l])
                    l++
            }
        }
    }
    return res;
};

// 思路首先是得先把数组排序
// 比如说是 i, j, n
// 先遍历数组 i 我们会选择跳过重复的值 j = i + 1, n = len - 1
// 然后再看 nums[i] + nums[j] + nums[n] 的值
// 大于0的话 n-- 小于0的话 j++ 
// 等于0的时候 得确定接下来的 nums[j] 和 nums[j - 1] 不同
// 这样才能确保是每个数组都是不同的值

let nums = [-1,0,1,2,-1,-4];
let nums1 = [0,1,1];
let nums2 = [0,0,0];
console.log(threeSum(nums));
console.log(threeSum(nums1));
console.log(threeSum(nums2));

File: specified-learning专项训练/slidingWindowAndTwoPointers/twoPointers/exercices/15-3Sum.py
# 15. 3Sum
# Given an integer array nums, return all the triplets [nums[i], nums[j], nums[k]] such that i != j, i != k, and j != k, and nums[i] + nums[j] + nums[k] == 0.
# Notice that the solution set must not contain duplicate triplets.

from typing import List
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        i = 0
        while i < len(nums):
            l = i + 1
            r = len(nums) - 1
            while l < r:
                tmpSum = nums[i] + nums[l] + nums[r]
                if tmpSum == 0:
                    res.append([nums[i], nums[l], nums[r]])
                    l = l + 1
                    while l < len(nums) and nums[l] == nums[l - 1]:
                        l = l + 1
                elif tmpSum > 0:
                    r = r - 1
                else:
                    l = l + 1
            i = i + 1
            while i < len(nums) and nums[i] == nums[i - 1]:
                i = i + 1
        return res

nums = [-1,0,1,2,-1,-4]
print(Solution().threeSum(nums))
nums = [0,1,1]
print(Solution().threeSum(nums))
nums = [0,0,0]
print(Solution().threeSum(nums))
